### DISCUSSION TITLE
Why this error is occurring  "attempted to change the reference count of a Python object while the GIL"

### DISCUSSION THREAD

**Discussion by matrixbot123** [wjakob/nanobind#238] [Q&A]
So , I am trying to bind HPX library async function but there is issue with calling it in recursive function . I dont understant it because I am not changing the referece counting manually but for some reason its showing usp and I dont believe HPX will do that
So for example : simple fibonacci funtion 
```py
import pyhpx as p
p.init_hpx()
def fibonacci(n):
    if n<=1:
        return 1
    n1= p.new_async(fibonacci,n-1)
    n2= p.new_async(fibonacci,n-2)
    
    return  n2.get()+n1.get()
print(fibonacci(10))
```
this is the output it gives https://gist.github.com/matrixbot123/82e993bf8b5296b5de8a6af6b44099ae
you can see there this error message
 ```
nanobind::detail::decref_check(): attempted to change the reference count of a Python object while the GIL was not held
```
but I have can done that in my bindings . can someone tell whyis that happening 

For reference these are the binding 
```cpp
void init_hpx_runtime(){
    // nb::gil_scoped_release release;
    // nb::gil_scoped_acquire acquire;
    int argc = 0;
    char** argv = nullptr;
    hpx::start(nullptr, argc, argv);
}

int get(const hpx::shared_future<int>& future){
    nb::gil_scoped_release release;
    return future.get();
}

//async function that takes function from 
hpx::shared_future<int> async_function(std::function<int(int)> f, int a)
{   std::cout<<"called async "<<a<<std::endl;
    nb::gil_scoped_release release;
    return hpx::async([f , a]() {
        nb::gil_scoped_acquire acquire;
        return f(a);        
    });
}

NB_MODULE(pyhpx , m){ 
    m.def(  "init_hpx",
            &init_hpx_runtime);
                     
    nb::class_ <hpx::shared_future<int>>(m,"shared_futureInt")
        .def(nb::init<>())
        .def(nb::init<const hpx::shared_future<int>&>())
        // .def("set_value", &set_future_value)
        // .def("is_ready",&is_ready,"Checks if the associated shared state is ready")
        .def("get",&get,"Get the value of shared future")
    ;

    m.def("new_async", 
        [](std::function<int(int)> f,int a){return async_function(f,a); }
        
    );

}
```

**Comment by wjakob** [comment]
Does this "HPX" library spawn threads? If so, there is your problem. 

It's fine to use nanobind API from custom threads, but you need to acquire the GIL first before you do. I suggest compiling in debug mode and setting a breakpoint on "abort".  That should give you a backtrace indicating the problem.

**Reply by matrixbot123** [reply]
Why is it required the acquire gil if it is not releated to python object. How to debug it when you have to an .so file and python script ?
